### PR TITLE
Windows: Use FILE_ALLOCATION_INFO for truncation

### PR DESCRIPTION
**Pull request by ChrisDenton**
We use `FILE_END_OF_FILE_INFO` here only because WINE does not support `FILE_ALLOCATION_INFO`. Instead of going with the one with broadest support, let's just use that as fallback only.

### CODE REVIEW

**Comment by Amanieu** [comment]
What is the benefit of using `FILE_ALLOCATION_INFO`? Does `FILE_END_OF_FILE_INFO` not actually reduce the allocated size on disk?

**Comment by ChrisDenton** [comment]
Tbh, I've been unable to find a difference in the case of `0`. In either case the reported allocation beyond EOF is freed when the handle is closed (and not before).

My main reason for doing this is because it's what Windows generally uses to truncate a file which makes me think there's some reason I don't know. Maybe there is a filesystem driver that does something differently.

**Comment by ChrisDenton** [comment]
Ah, setting `FILE_END_OF_FILE_INFO` *should* imply also setting the allocation size. But that's not a *must*.

> If InputBuffer.EndOfFile is less than (BlockAlign(Open.Stream.Size, Open.File.Volume.ClusterSize) -Open.File.Volume.ClusterSize), the object store SHOULD set Open.Stream.AllocationSize to BlockAlign (InputBuffer.EndOfFile, Open.File.Volume.ClusterSize).

And a difference between the two is for setting the allocation:

> If the object store supports Open.File.Volume.ClusterRefcount, for each EXTENTS that is removed from Open.Stream.ExtentList as a result of truncation, for each cluster that is being referred to by the EXTENTS being removed, its entry in Open.File.Volume.ClusterRefcount MUST be decremented. If the corresponding cluster's reference count goes to zero, then that cluster MUST also be freed.

**Comment by Amanieu** [comment]
Since this is what Windows does anyways there's no harm in us doing the same.

File: library/std/src/sys/pal/windows/fs.rs [hunk 1]
```diff
@@ -316,19 +316,31 @@ impl File {
...
-                    // better so let's just use the more widely supported method.
-                    let eof = c::FILE_END_OF_FILE_INFO { EndOfFile: 0 };
+                    // This first tries `FileAllocationInfo` but falls back to
+                    // `FileEndOfFileInfo` in order to support WINE.
+                    // If WINE gains support for FileAllocationInfo, we should
+                    // remove the fallback.
+                    let alloc = c::FILE_ALLOCATION_INFO { AllocationSize: 0 };
                     let result = c::SetFileInformationByHandle(
                         handle.as_raw_handle(),
                         c::FileEndOfFileInfo,
```
[the remark this answers is not in the document]
Comment (ChrisDenton) [reply library/std/src/sys/pal/windows/fs.rs:326]: Oh, huh. I fixed that locally but somehow forgot to push. Fortunately they are the same size (and if they weren't then it'd return an error).

**Pull request merged**